### ISSUE TITLE
Log config to wandb

### ISSUE THREAD

**Issue by levan92** [open-mmlab/mmcv#1773]
Although there is `WandbLoggerHook`, but the config of the run is not logged to wandb. Will be nice if this can be done by default as long as `WandbLoggerHook` is invoked. 

Currently, this is what I'm adding to my training script in order to log the config:  

```python
wandb_kwargs = [ h['init_kwargs'] for h in cfg.log_config.hooks if h['type'] == 'WandbLoggerHook' ][0]
wandb_kwargs['config'] = cfg._cfg_dict.to_dict()
wandb.init(**wandb_kwargs)
logger.info('Logged cfg to wandb.') 
```

Is there any way to get access to the cfg object from `WandbLoggerHook`? I'll be happy to put in a PR for this if needed. 

Thank you!

**Comment by teamwong111** [comment]
It seems that there is no way to get access to the cfg object from `WandbLoggerHook`.
Maybe you can modify cfg object in `train.py` like this
```
for h in cfg.log_config.hooks:
    if h['type'] == 'WandbLoggerHook':
        h['init_kwargs']['config'] = cfg._cfg_dict.to_dict()
```

Do you have any idea about this issue?

**Comment by levan92** [comment]
Yup, I'm already doing that in `train.py`. If there's no way to get access to cfg in a hook then I guess there's no choice.

**Comment by ayulockin** [comment]
I am trying to achieve the same thing. While investigating, I realized that the `WandbLoggerHook` methods have access to the `runner` object. However, the `runner` has no `cfg` class variable. 

If the constructor of `BaseRunner` can accept `cfg` then the `WandbLoggerHook` (and other hooks) can use it to log the config file, something like this:

```
def before_run(self, runner):
        super(WandbLogger, self).before_run(runner)
        self.wandb.config = runner.cfg
```

Something as simple as this, should do the trick. 

```
class BaseRunner(metaclass=ABCMeta):
       ....
      def __init__(self, ...., cfg=None):
           self.cfg = cfg
```

In the `train.py` we can then do `runner.cfg = cfg`. 

Thoughts? If this is an acceptable solution, I can make a PR.

**Comment by ayulockin** [comment]
For now, I am doing this to log config to W&B:

```
cfg.log_config.hooks = [
    dict(type='WandbLogger',
         init_kwargs={'entity': 'ayut', 'project': 'mmdetection-test', 'config': cfg._cfg_dict.to_dict()},
         interval=10,
         ignore_last=True,
         reset_flag=False,
         commit=True,
         by_epoch=True,
         with_step=True,
         log_artifact=False)]
```

<img width="594" alt="image" src="https://user-images.githubusercontent.com/31141479/157172660-cba9a52b-114b-44d2-89e9-987b4b47e171.png">

**Comment by teamwong111** [comment]
> I am trying to achieve the same thing. While investigating, I realized that the `WandbLoggerHook` methods have access to the `runner` object. However, the `runner` has no `cfg` class variable.
> 
> If the constructor of `BaseRunner` can accept `cfg` then the `WandbLoggerHook` (and other hooks) can use it to log the config file, something like this:
> 
> ```
> def before_run(self, runner):
>         super(WandbLogger, self).before_run(runner)
>         self.wandb.config = runner.cfg
> ```
> 
> Something as simple as this, should do the trick.
> 
> ```
> class BaseRunner(metaclass=ABCMeta):
>        ....
>       def __init__(self, ...., cfg=None):
>            self.cfg = cfg
> ```
> 
> In the `train.py` we can then do `runner.cfg = cfg`.
> 
> Thoughts? If this is an acceptable solution, I can make a PR.

@zhouzaida , any opinion about this?

**Comment by ayulockin** [comment]
Hey @zhouzaida, would love to know what you think about it. I can make a PR if you would like.
